drivers/hid/hidclass/fdo.c

```diff
@@ -388,6 +390,215 @@ HidClassFDO_GetDescriptors(
+        {
+            // find PDO related to the reportID
+            {
+                CollectionNumber = HidClassPDO_GetReportDescriptionByReportID(&DeviceExtension->Common.DeviceDescription, DeviceExtension->InputBuffer[0])->CollectionNumber;
+
+                for (int i = 0; i < DeviceExtension->DeviceRelations->Count; i++)
+                {
+                    PDODeviceExtension = DeviceExtension->DeviceRelations->Objects[i]->DeviceExtension;
+                    ASSERT(PDODeviceExtension->Common.IsFDO == FALSE);
+
+                    if (PDODeviceExtension->CollectionNumber == CollectionNumber)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        PDODeviceExtension = NULL;
+                    }
+                }
+            }
+
+            ASSERT(PDODeviceExtension != NULL);
+
+
+            InitializeListHead(&completeIRP);
+
+            KeAcquireSpinLock(&PDODeviceExtension->ReadLock, &CurrentIRQL);
+
+            // check if any IRP waiting data
+            if (IsListEmpty(&PDODeviceExtension->PendingIRPList) == FALSE)
+            {
+                PUCHAR address = NULL;
+                PIRP tempIRP = NULL;
+
+                while (!IsListEmpty(&PDODeviceExtension->PendingIRPList))
+                {
+                    tempIRP = HidClassPDO_DequeueIrp(PDODeviceExtension);
+                    if (tempIRP != NULL)
+                    {
+                        if (tempIRP->MdlAddress != NULL)
+                        {
+                            address = MmGetSystemAddressForMdlSafe(tempIRP->MdlAddress, LowPagePriority);
+                            if (address != NULL)
+                            {
+                                tempIRP->IoStatus.Information = HidClassPDO_GetCollectionDescription(&PDODeviceExtension->Common.DeviceDescription, PDODeviceExtension->CollectionNumber)->InputLength;
+                                RtlCopyMemory(address, DeviceExtension->InputBuffer, tempIRP->IoStatus.Information);
+                                tempIRP->IoStatus.Status = STATUS_SUCCESS;
+                            }
+                            else
+                            {
+                                tempIRP->IoStatus.Status = STATUS_INVALID_USER_BUFFER;
+                                tempIRP->IoStatus.Information = 0;
+                            }
+                        }
+
+                        // Mark this IRP to be cenceled after exiting spinLock
+                        InsertHeadList(&completeIRP, &tempIRP->Tail.Overlay.ListEntry);
+                    }
+                }
+            }
+
+            // No IRP waiting data, put it into the buffer
+            if (IsListEmpty(&completeIRP) == TRUE)
+            {
+                HidClass_CyclicBufferPut(&PDODeviceExtension->InputBuffer, DeviceExtension->InputBuffer);
```

Comment:
I don't quite get the idea - why do you need a cyclic buffer here? It feels like only a single element is always used from there

Reply:
it's usual for applications like games, to perform input read after\before own calculations. so in these situations driver must store input reports. see hidd_setnuminputbuffers